### PR TITLE
At-least-once API

### PR DESCRIPTION
**Pull request by xhl1988**

### CODE REVIEW

File: lib/kafka_base_producer.js [hunk 1]
```diff
@@ -0,0 +1,612 @@
...
+        }
+        return;
+    }
+
+    if (self.restClient) {
+        var produceMessage = self.getProduceMessage(topic, producerRecord.key,
+            producerRecord.value, Date.now() / 1000, 'binary');
+        self._produce(produceMessage, callback);
+
+        if (self.enableAuditC3 && topic !== self.auditTopicNameC3) {
```
Comment (cloudXane) [inline lib/kafka_base_producer.js]: better limit getTsFromHp for hp topics

**Comment by mmchen** [comment]
lgtm

**Comment by mmchen** [comment]
looks good in general, please resolve conflict before merging

File: lib/kafka_rest_client.js [hunk 2]
```diff
@@ -211,7 +211,7 @@ KafkaRestClient.prototype.isHeatPipeTopic = function isHeatPipeTopic(topic) {
 KafkaRestClient.prototype.goToLocalAgent = function goToLocalAgent(topic) {
     var self = this;
     return (self.clientType !== 'AtLeastOnce' && self.isHeatPipeTopic(topic)) ||
-        (self.clientType === 'AtLeastOnce' && self.localAgentMode === 'Always');
+        (self.clientType === 'AtLeastOnce' && (self.localAgentMode === 'Always' || self.localAgentMode === 'ALWAYS'));
```
Comment (xhl1988) [inline lib/kafka_rest_client.js]: Yes. I'll use toUpperCase()

**Pull request merged**